### PR TITLE
Remove duplicate strong-named assemblies from modules supplied by the Debugger

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@tmat, @amcasey, @KevinH-MS, @ManishJayaswal, @pnelsonmsft

**Comment by amcasey** [comment]
Can you please provide a high-level overview of the change?  I'm assuming this is to address the problems we've been having with loading two copies of the same assembly (but not two copies of the same type from different assemblies).

**Comment by tmat** [comment]
I'm not sure if the effect of this change is entirely desirable. Semantic versioning in nuget world allows breaking changes between revs as long as the major version is bumped. So class C, v1.0 can be incompatible with C, v2.0. Both versions can be loaded in the same process when used by different components. EE would be broken in the component that uses v1.0. 

It seems we need to be smarter than just picking the highest version. We are evaluating in a context of a frame that belongs to some assembly. If that assembly references v1.0 we should use that module. There might still be conflicts if the assembly indirectly references v1.0 and v2.0. So we might need to get even more sophisticated. 

We can leave this to Update 1 though.

**Comment by cston** [comment]
Overview: Handle cases where the same assembly is loaded multiple times in debuggee, with same or different version. This can arise for example if the debuggee explicitly loads an assembly at runtime that the exe also references.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 1]
```diff
@@ -113,6 +113,71 @@ internal static ImmutableArray<MetadataReference> MakeAssemblyReferences(this Im
...
+                    continue;
+                }
+                var bestIndex = i;
+                for (int j = i + 1; j < n; j++)
+                {
+                    if (!IsNonNullAndStrongNamed(identities[j]))
+                    {
+                        continue;
+                    }
+                    var compareResult = AssemblyIdentityComparer.Default.Compare(identities[bestIndex], identities[j]);
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Should the AssemblyIdentityComparer be a parameter?  I think the compilation (options?) has one.

**Comment by amcasey** [comment]
@tmat Do we also need to consider cases where the compile time reference was to version X, but versions Y and Z are available during debugging?

**Comment by amcasey** [comment]
I'm not convinced that it is always safe to keep only the higher version, but at least it's easy to explain what we're doing.  If someone is legitimately referencing two versions and we break their experience by dropping the older one, it will be very easy to explain to them when they can expect to have problems.

**Comment by tmat** [comment]
Yeah, this is gonna be non-trivial.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 2]
```diff
@@ -437,94 +436,7 @@ internal string GetFullNameOrThrow(Handle namespaceHandle, StringHandle nameHand
         /// <exception cref="BadImageFormatException">An exception from metadata reader.</exception>
         internal AssemblyIdentity ReadAssemblyIdentityOrThrow()
```
Comment (VladimirReshetnikov) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: This method is now trivial. Should we just inline it?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 3]
```diff
@@ -3811,95 +3811,72 @@ .maxstack  1
...
+                assemblyName: ExpressionCompilerUtilities.GenerateUniqueName());
+            var referenceA = AssemblyMetadata.CreateFromImage(compilationA.EmitToArray()).GetReference();
+            var identityA = referenceA.GetAssemblyIdentity();
+
             var compilationB = CreateCompilationWithMscorlib(
                 sourceB,
                 options: TestOptions.DebugDll,
-                assemblyName: assemblyName + "_B",
+                assemblyName: ExpressionCompilerUtilities.GenerateUniqueName(),
                 references: new MetadataReference[] { referenceA });
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: `new[] { referenceA }`

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 4]
```diff
@@ -3811,95 +3811,72 @@ .maxstack  1
...
+                ImmutableArray.Create(systemRefIdentity, mscorlibIdentity));
+            // Strong-named duplicate, lower version first.
+            VerifyAssemblyReferences(
+                ImmutableArray.Create(SystemRef_v20, SystemRef, SystemRef_v20),
+                ImmutableArray.Create(systemRefIdentity));
+            // Strong-named duplicate, higher version first.
+            VerifyAssemblyReferences(
+                ImmutableArray.Create(SystemRef, SystemRef_v20, SystemRef_v20),
+                ImmutableArray.Create(systemRefIdentity));
+            // Strong-named and non-strong named duplicates.
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: "non-strong-named"

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 5]
```diff
@@ -113,6 +113,71 @@ internal static ImmutableArray<MetadataReference> MakeAssemblyReferences(this Im
             return referencesBuilder.ToImmutableAndFree();
         }
 
+        private static void RemoveStrongNamedDuplicates(ArrayBuilder<MetadataReference> references, ArrayBuilder<AssemblyIdentity> identities)
+        {
+            Debug.Assert(references.Count == identities.Count);
+            Debug.Assert(identities.All(i => i != null));
+
+            // Null out any identities that are duplicates. Note, this is O(n^2).
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Could be done in O(n log n), if you sort them first. Not sure if it worth it.

**Comment by KevinH-MS** [comment]
Is this consistent with what the old EE did?  While I agree with @tmat that we should try to resolve against the assembly directly referenced (by the assembly containing the current instruction address), if we're at least consistent with the old implementation, I think we're good (at least for now).

**Comment by cston** [comment]
Updated to attach extern aliases to duplicate assemblies for assemblies that match references within the target module only but have different version.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 6]
```diff
@@ -110,6 +121,60 @@ internal static ImmutableArray<MetadataReference> MakeAssemblyReferences(this Im
...
+                    duplicateIdentities[identity] = Guid.NewGuid().ToString();
+                }
+            }
+
+            Debug.Assert(references.All(r => r.Properties.Aliases.IsEmpty));
+
+            // Add extern aliases to duplicate references.
+            if (duplicateIdentities.Count > 0)
+            {
+                for (int i = 0; i < n; i++)
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: I think I'm missing the purpose of the second pass through identities.  Can this just be inlined into the loop above (and get rid of the Dictionary)?

Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Never mind...

Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Sorry, no...question stands...can this be inlined?

**Comment by KevinH-MS** [comment]
LGTM :shipit:

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 6 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: The Dictionary allows attaching the same extern alias to each duplicate in the (unlikely) event there are multiple duplicates with the same identity, and ensures we attach at most one extern alias in the (also unlikely) event that an assembly is a duplicate of multiple references. In short, the Dictionary may not be strictly necessary.

Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: I see--thanks for explaining. :-)

Consider: a comment to the above effect

**Comment by cston** [comment]
Note, VB does not support extern aliases so even though aliases are added to duplicate modules, the types are still available in the EE. In other words, references to types in duplicate modules will be ambiguous in the EE. The VB test has been updated to reflect that.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: There are a number of uses of this method so we should probably leave as is.

**Comment by amcasey** [comment]
I'm not sure I see the benefit of a change that makes things better in some C# cases, worse in some C# cases, and has no effect at all on VB.  Am I missing something?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 7]
```diff
@@ -110,6 +121,62 @@ internal static ImmutableArray<MetadataReference> MakeAssemblyReferences(this Im
...
+                    // ReferenceManager will drop equivalent assemblies.
+                    if (compareResult != AssemblyIdentityComparer.ComparisonResult.EquivalentIgnoringVersion)
+                    {
+                        continue;
+                    }
+                    duplicateIdentities[identity] = Guid.NewGuid().ToString();
+                }
+            }
+
+            Debug.Assert(references.All(r => r.Properties.Aliases.IsEmpty));
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Why is this true?  Does this code run before we re-constitute extern aliases from custom debug info?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MethodContextReuseConstraints.cs [hunk 8]
```diff
@@ -9,40 +9,31 @@ namespace Microsoft.CodeAnalysis.ExpressionEvaluator
 {
     internal struct MethodContextReuseConstraints
     {
-        private readonly Guid _moduleVersionId;
```
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MethodContextReuseConstraints.cs]: Why remove the MVID?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 7 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Yes, this runs before looking at the PDBs.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MethodContextReuseConstraints.cs (hunk 8 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MethodContextReuseConstraints.cs]: We're now checking the MVID in *MetadataContext.Matches.

**Comment by cston** [comment]
@amcasey Are there C# cases where this change makes the behavior worse?

**Comment by cston** [comment]
@amcasey Are there C# cases where this change makes the behavior worse? The behavior for C# should be reasonable in all cases.

**Comment by jasonwilliams200OK** [comment]
This PR was replaced by #1687. Probably a good idea to close this one (and also the other old/stale PRs and issues)? :)
Roslyn repo needs some cleanup from this perspective.

**Pull request closed without merging**